### ISSUE TITLE
Add an optional argument for the "overlay_XXX" functions(former refFunctions)

### ISSUE THREAD

**Issue by PQuesnot** [qgis/QGIS#44652]
### Feature description

Hello everyone,

Before I start, I am not a developer, so everything I am going to say is based on my personal feeling as a QGIS user.
Also, I am French and my English is a bit weak, so please be patient.

the "overlay_XXX" functions have been introduced with the release of QGIS 3.16.
These functions are very useful to make spatial joins and spatial conditions. 
However, in my opinion, there is an option missing that could make these new functions even more powerful.

From what I understand, these formulas compare the geometry of a "source" layer with the geometry of a "target" layer and return different values depending on the arguments used.

However, it is not possible to specify the "source" geometry.
Let's take an example: 

I have a "line" layer and a "node" layer, 
the "node" layer has only one attribute --> "data
the "line" layer has two attributes --> "node_up_data" and "node_down_data

for the "node_up_data" attribute I want to get the "data" information of the point intersecting the first point of my line
and for the attribute "node_down_data" I want to get the data of the point intersecting the last point of my line

so I use the formulas overlay_intersects (by putting an index of 0 and 1)
and the results are not the expected ones, they are even totally random

I propose to add an optional argument in the formulas "overlay_XXX" to specify the "source" geometry (by default equal to $geometry)

example of syntax : 
overlay_intersects(layer[,expression][,filter][,limit][,cache=false][,source_geom=$geometry])

for my previous example, the formulas become : 
overlay_intersects(layer:='node',expression:="data",source_geom:=start_point($geometry))
and 
overlay_intersects(layer:='node',expression:="data",source_geom:=end_point($geometry))

So here is my idea, 
what do you think ? 

### Additional context

![Capture](https://user-images.githubusercontent.com/36067581/129011659-1009508f-a9ec-44be-b27b-3dd22ad943c8.PNG)

**Comment by kannes** [comment]
If I understand correctly this is possible via `geometry(@parent)`

**Comment by agiudiceandrea** [comment]
> If I understand correctly this is possible via `geometry(@parent)`

@kannes, I don't think this could solve the issue. With the overlay_* functions there is no parameter to specify any geometry other than the source geometry, so AFAIK you cannot use `geometry(@parent)` in any place.

**Comment by kannes** [comment]
Yes, I think I misunderstood the problem. :)

I had `aggregate` in my head where something like this might be an alternative approach:

```
aggregate(
  layer:='node',
  aggregate:='concatenate',
  expression:="data",
  filter:=intersects(
    $geometry,
    start_point(geometry(@parent))  -- if the parent layer is the one with the lines
  )
)
```

**Comment by agiudiceandrea** [comment]
@gioman. I think this is a valid feature request and the "feedback" tag should be removed.
